### PR TITLE
Onboarding Project: App-Cleanup

### LINKED ISSUE #47: General App Clean-up
## Main Goal
Make your app seem less like a onboarding project and more like and interface used by... not us! Basically you're going to make it a more intuitive interface rather than relying on a navigation bar to get from page to page. 

### When Accomplished
- [x] The user list is the home page.
  - [x] When there is no users in the user list, the home page should prompt the user to create a user
  - [x] Somehow links to, or has the ability to, add a new user
    -  _If linked, the app routes back to the user list on completion of creation_
- [x] Navigation bar is no more
  - [x] User profile and user list pages are no longer apart of the project
  - [x] The nav component is no longer apart of the project

### PR DESCRIPTION
**Pull request by Jropp**
## What It Does

The app is now consolidated into a single page. It opens to the list of users. From there you can edit existing users or create new users.

The former design was less intuitive and required clicking a link and static page loads to switch between creating a new user and seeing the list of users.

The users-list component became the parent component. I used dom-if templates and a modes property to set the state of the app.  Clicking on 'create new user', 'save', 'cancel', etc alters the state of the app rather than making page changes.

Closes #47 

## How To Test

1. Create a new user and save. It should display immediately in the user-list
2. Edit a user's data and save. It should display immediately.
3. Delete all users. You should be prompted that the database is empty and to create a new user. 
4. Create a new user but cancel instead of saving. You should see the user display page again.
5. When there are no users in the database, start filling out the new user fields. Click cancel. The fields should be emptied out. 

## Notes/Caveats

This still implements local storage for the database. This will be changed in #48 

## Checklist

Under penalty of public shaming, I avow that I:

- [ ] Reviewed the diff to look for typos, whitespace errors, and console.log() statements
- [ ] Verified that there are no compiler or linter errors or warnings
- [ ] Verified the build in production(optimized) mode
- [ ] Updated the docs, if necessary
- [ ] Included the appropriate labels
- [ ] Referenced applicable issues (i.e. Closes #XXXX, Fixes #XXXX)

Browsers tested:
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge
- [ ] Internet Explorer

## Dependencies

none

### CODE REVIEW

File: src/user-list/user-list.html [hunk 1]
```diff
@@ -41,35 +41,73 @@
...
+                DISPLAY: 'display'
+              };
+            }
+          },
         };
       }
+      connectedCallback() {
+        super.connectedCallback();
+        if (!this.users.length) {
+          this.users = false;
```
Comment (kddrake) [inline src/user-list/user-list.html]: `this.users` is of type `Array`... why set it to `false`?

File: src/user-list/user-list.html [hunk 2]
```diff
@@ -41,35 +41,73 @@
...
+    <template is="dom-if" if="[[isMode(mode, modes.DISPLAY)]]">
+      <button on-click="createNewUserCard">+ Create New User</button>
+    </template>
+    <template is="dom-if" if="[[isMode(mode, modes.CREATE_NEW)]]">
+      <template is="dom-if" if="[[!users]]">
+        <h2>There are currently no users in the database. Please create one.</h2>
+      </template>
+      <user-component mode="edit"></user-component>
+    </template>
+    <template is="dom-if" if="[[isMode(mode, modes.DISPLAY)]]">
```
Comment (kddrake) [inline src/user-list/user-list.html]: Is it necessary to have two `dom-if`s with the same `if` condition?

File: src/index.html [hunk 3]
```diff
@@ -6,7 +6,7 @@
 	<meta http-equiv="X-UA-Compatible" content="IE=edge">
 	<title>Janky Homepage</title>
```
[the remark this answers is not in the document]
Comment (Jropp) [reply src/index.html]: x

File: src/user-list/user-list.html [hunk 4]
```diff
@@ -72,6 +92,10 @@ <h2>There are currently no users in the database. Please create one.</h2>
             type: Array,
             value: () => JSON.parse(localStorage.getItem('onboardProjectUsers')) || []
           },
+          usersToDisplay: {
+            type: Boolean,
+            value: true,
```
Comment (kddrake) [inline src/user-list/user-list.html]: I'm not sure you'd want to default to be `true`. I think it makes more sense to set this to `true` once you're positive there are users.

File: src/index.html [hunk 5]
```diff
@@ -4,10 +4,9 @@
 <head>
 	<meta charset="utf-8" />
 	<meta http-equiv="X-UA-Compatible" content="IE=edge">
-	<title>Janky Homepage</title>
+	<title>Database Users</title>
```
Comment (kddrake) [inline src/index.html]: Let's go with something like 'Jropp IOP' or 'Jason Onboarding' or something to do with your name and the fact that this is your project.

File: src/user-list/user-list.html [hunk 6]
```diff
@@ -44,22 +44,42 @@
...
+          </user-component>
+        </template>
+      </div>
+      <div class="new-user-column">
+        <div class="create-user-component">
+          <template is="dom-if" if="[[isMode(mode, modes.DISPLAY)]]">
+            <button on-click="createNewUserCard">+ Create New User</button>
+          </template>
+          <template is="dom-if" if="[[isMode(mode, modes.CREATE_NEW)]]">
+            <template is="dom-if" if="[[!usersToDisplay]]">
```
Comment (kddrake) [inline src/user-list/user-list.html]: I think this "empty" message makes more sense in the 'user-list-column' than it does with in the 'new-user-column' as it's signifying the list is empty.

File: src/user-list/user-list.html (hunk 4 above)
Comment (Jropp) [reply src/user-list/user-list.html]: x

File: src/index.html (hunk 5 above)
Comment (Jropp) [reply src/index.html]: x

**Comment by Jropp** [comment]
@kddrake I moved the data management into user-list.html and it is currently working so that edit/save/delete of an existing user-card doesn't mess with the create new user card.

The thing I can't figure out is: 

If I am editing an existing user card and then save a new user before I have saved the existing user, the existing user will save the changes I've made rather than just leaving that component in edit mode.

**Comment by Jropp** [comment]
@kddrake if you could take a look at the functionality and make sure this matches the criteria we talked about.

Essentially what I did was set the user component up with two ways it can get blocked 
```
disableSave: {
						type: Boolean,
						value: false
					},
					disableEdit: {
						type: Boolean,
						value: false
					}
```

Whenever you click any edit button, it sends a disableFunctionality 'virus' to all components `<user-component disableSave/disableEdit="[[disableFunctionality]]">`
```
<template is="dom-repeat" items="[[users]]" as="user">
					<user-component disable-edit=[[disableFunctionality]] mode="display" user="[[user]]"></user-component>
				</template>
```
```
<template is="dom-if" if="[[isMode(mode, modes.CREATE_NEW)]]">
						<user-component disable-save=[[disableFunctionality]] mode="edit"></user-component>
					</template>
```

So the 'virus' only disables the edit functionality of items in users list and the save functionality of a component that is trying to create a new user. The virus is cured by hitting save, which is only possible in cards in the user-list who have had their edit functionality disabled but not the save, but is not possible in the new user card which has had its save functionality disabled.

**Comment by kddrake** [comment]
@collincahill Yeah that works for me 👍

File: src/user-list/user-list.html [hunk 7]
```diff
@@ -41,35 +41,154 @@
...
+          this.save(e.detail.user);
+          this.disableFunctionality = false;
+        });
+
+        document.addEventListener('edit', () => {
+          this.disableFunctionality = true;
+        });
+
+      }
+
```
Comment (kddrake) [inline src/user-list/user-list.html]: You should also have a `disconnectedCallback` [removing](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/removeEventListener) each event listener.

Comment (Jropp) [reply src/user-list/user-list.html]: x

File: src/user-component/user-component.html [hunk 8]
```diff
@@ -208,8 +231,21 @@ <h1>[[headerMessage(mode)]]</h1>
...
+            return;
+          }
+
+          this.dispatchEvent(new CustomEvent('edit', {
+            bubbles: true,
+            composed: true
+          }));
+        }
+
+        this.mode = this.isMode(this.mode, this.modes.EDIT) ? this.modes.DISPLAY : this.modes.EDIT;
```
Comment (kddrake) [inline src/user-component/user-component.html]: This is a bit convoluted as to what it's doing and more or less breaks "do one thing". I would just pull everything from the `if (e)` into it's own function that calls `toggleMode`.

Comment (Jropp) [reply src/user-component/user-component.html]: x

**Pull request merged**